homeassistant/components/powerwall/__init__.py

```diff
@@ -121,20 +122,17 @@ def _update_data(self) -> PowerwallData:
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Tesla Powerwall from a config entry."""
-    http_session = requests.Session()
     ip_address: str = entry.data[CONF_IP_ADDRESS]
 
     password: str | None = entry.data.get(CONF_PASSWORD)
-    power_wall = Powerwall(ip_address, http_session=http_session)
+    power_wall = Powerwall(ip_address)
```

Comment:
We should pass in the aiohttp session from the helper

Reply:
`websession = async_get_clientsession(hass, verify_ssl=False)`

Reply:
> `websession = async_get_clientsession(hass, verify_ssl=False)`

Thank you, I didn't know about that.    Is the session meant to be closed by this code then?   The `Powerwall` class does have a [close method](https://github.com/jrester/tesla_powerwall/blob/master/tesla_powerwall/powerwall.py#L192) that should really be called in these error cases.   As of now, it calls close on its session.

Reply:
If the session is passed in, the lib should not close it. It should only close sessions that it creates (is the owner of)

Reply:
> If the session is passed in, the lib should not close it. It should only close sessions that it creates (is the owner of)

In a related question, would it be a bug in the library [on this line](https://github.com/jrester/tesla_powerwall/blob/master/tesla_powerwall/api.py#L28) since it appears to override the session's value for using the default SSL validation?

Reply:
> > If the session is passed in, the lib should not close it. It should only close sessions that it creates (is the owner of)
> 
> 
> 
> In a related question, would it be a bug in the library [on this line](https://github.com/jrester/tesla_powerwall/blob/master/tesla_powerwall/api.py#L28) since it appears to override the session's value for using the default SSL validation?

Probably but since that session doesn't get reused elsewhere in HA it likely wasn't a problem

Reply:
There's an issue with `aiohttp.ClientSession` where it will reject cookies from URLs that contain IP addresses instead of DNS names.   See [the docs](https://docs.aiohttp.org/en/latest/client_advanced.html#cookie-jar).

Seems like this would hit a lot of folks since I figure most would use their gateway's IP address and not bother setting up any local naming service on their home network.

The workaround is to change how the session is created, but it makes it no longer compatible with `async_get_clientsession` since it doesn't allow us to pass kwargs to the sessions's constructor.   `async_create_clientsession` on the other hand does allow this, but it has the side effect of no longer being stored in `hass.data[DATA_CLIENTSESSION]`.

I wondered if you had any thoughts on the matter.   Seems like there isn't a good choice other than to use `async_create_clientsession` and lose out on potential session caching.

Reply:
I think what you are looking for is

https://github.com/home-assistant/core/blob/4e62dacc00fa2596eb94e66f053af8882363839f/homeassistant/components/isy994/__init__.py#L79

Reply:
Yes exactly!  :)  Too bad it doesn't allow it to be stored in the hass data [like it does here](https://github.com/home-assistant/core/blob/4e62dacc00fa2596eb94e66f053af8882363839f/homeassistant/helpers/aiohttp_client.py#L86).

Part of this PR is that I removed `POWERWALL_HTTP_SESSION` which looked like something no longer used.   Maybe it would be good to store the session back in `hass.data` using that old key?